op-node/rollup/derive/engine_update.go

```diff
@@ -117,21 +118,37 @@ func startPayload(ctx context.Context, eng ExecEngine, fc eth.ForkchoiceState, a
 // confirmPayload ends an execution payload building process in the provided Engine, and persists the payload as the canonical head.
 // If updateSafe is true, then the payload will also be recognized as safe-head at the same time.
 // The severity of the error is distinguished to determine whether the payload was valid and can become canonical.
-func confirmPayload(ctx context.Context, log log.Logger, eng ExecEngine, fc eth.ForkchoiceState, id eth.PayloadID, updateSafe bool) (out *eth.ExecutionPayload, errTyp BlockInsertionErrType, err error) {
-	payload, err := eng.GetPayload(ctx, id)
+func confirmPayload(ctx context.Context, log log.Logger, eng ExecEngine, fc eth.ForkchoiceState, id eth.PayloadID, updateSafe bool, agossip async.AsyncGossiper) (out *eth.ExecutionPayload, errTyp BlockInsertionErrType, err error) {
+	var payload *eth.ExecutionPayload
+	// if the payload is available from the async gossiper, it means it was not yet imported, so we reuse it
+	if cachedPayload := agossip.Get(); cachedPayload != nil {
+		payload = cachedPayload
+		// log a limited amount of information about the reused payload, more detailed logging happens later down
+		log.Debug("found uninserted payload from async gossiper, reusing it and bypassing engine",
+			"hash", payload.BlockHash,
+			"number", uint64(payload.BlockNumber),
+			"parent", payload.ParentHash,
+			"txs", len(payload.Transactions))
+	} else {
+		payload, err = eng.GetPayload(ctx, id)
+	}
 	if err != nil {
 		// even if it is an input-error (unknown payload ID), it is temporary, since we will re-attempt the full payload building, not just the retrieval of the payload.
 		return nil, BlockInsertTemporaryErr, fmt.Errorf("failed to get execution payload: %w", err)
 	}
 	if err := sanityCheckPayload(payload); err != nil {
 		return nil, BlockInsertPayloadErr, err
 	}
+	// begin gossiping as soon as possible
+	// agossip.Clear() will be called later if an non-temporary error is found, or if the payload is successfully inserted
+	agossip.Gossip(payload)
 
 	status, err := eng.NewPayload(ctx, payload)
 	if err != nil {
 		return nil, BlockInsertTemporaryErr, fmt.Errorf("failed to insert execution payload: %w", err)
 	}
 	if status.Status == eth.ExecutionInvalid || status.Status == eth.ExecutionInvalidBlockHash {
+		agossip.Clear()
 		return nil, BlockInsertPayloadErr, eth.NewPayloadErr(payload, status)
 	}
 	if status.Status != eth.ExecutionValid {
```

Comment:
Is it just eth.ExecutionSyncing which could trigger this codepath? I think that we are clearing correctly with the errors we are inspecting, but I wonder if there are more error types we are not exhaustively matching on.

Reply:
I went through the errors we return in `NewPayloadErr` and did my best to evaluate this, LMK if it makes sense:

At this point in the execution, we've already checked `ExecutionInvalid` and `ExecutionInvalidBlockHash`, so now we are in `!=ExecutionValid` space.

Looking at the `NewPayloadErr` function, ignoring the statuses we've already eliminated, the remaining possibilities are:
* ExecutionSyncing
* ExecutionInvalidTerminalBlock
* ExecutionAccepted
* Default/Unknown

And the handling of each seems ok too:
* ExecutionSyncing should not be possible for Sequencers
* ExecutionInvalidTerminalBlock, based on the message, appears to be a fundamentally unexpected situation (like, the sequencer would need to be fully broken to see this)
* ExecutionAccepted appears to be a happily retryable error
* Default/Unknown is... unknown!